python/ray/data/dataset_pipeline.py

```diff
@@ -700,6 +682,325 @@ def __iter__(self):
+                pandas_json_args_fn=pandas_json_args_fn,
+                ray_remote_args=ray_remote_args,
+                **pandas_json_args,
+            )
+        )
+
+    def write_csv(
+        self,
+        path: str,
+        *,
+        filesystem: Optional["pyarrow.fs.FileSystem"] = None,
+        try_create_dir: bool = True,
+        arrow_open_stream_args: Optional[Dict[str, Any]] = None,
+        block_path_provider: BlockWritePathProvider = DefaultBlockWritePathProvider(),
+        arrow_csv_args_fn: Callable[[], Dict[str, Any]] = lambda: {},
+        ray_remote_args: Dict[str, Any] = None,
+        **arrow_csv_args,
+    ) -> None:
+        """Call ``Dataset.write_csv`` on each output dataset of this pipeline."""
+        self._write_each_dataset(
+            lambda ds: ds.write_csv(
+                path,
+                filesystem=filesystem,
+                try_create_dir=try_create_dir,
+                arrow_open_stream_args=arrow_open_stream_args,
+                block_path_provider=block_path_provider,
+                arrow_csv_args_fn=arrow_csv_args_fn,
+                ray_remote_args=ray_remote_args,
+                **arrow_csv_args,
+            )
+        )
+
+    def write_parquet(
+        self,
+        path: str,
+        *,
+        filesystem: Optional["pyarrow.fs.FileSystem"] = None,
+        try_create_dir: bool = True,
+        arrow_open_stream_args: Optional[Dict[str, Any]] = None,
+        block_path_provider: BlockWritePathProvider = DefaultBlockWritePathProvider(),
+        arrow_parquet_args_fn: Callable[[], Dict[str, Any]] = lambda: {},
+        ray_remote_args: Dict[str, Any] = None,
+        **arrow_parquet_args,
+    ) -> None:
+        """Call ``Dataset.write_parquet`` on each output dataset of this pipeline."""
+        self._write_each_dataset(
+            lambda ds: ds.write_parquet(
+                path,
+                filesystem=filesystem,
+                try_create_dir=try_create_dir,
+                arrow_open_stream_args=arrow_open_stream_args,
+                block_path_provider=block_path_provider,
+                arrow_parquet_args_fn=arrow_parquet_args_fn,
+                ray_remote_args=ray_remote_args,
+                **arrow_parquet_args,
+            )
+        )
+
+    def write_datasource(
+        self,
+        datasource: Datasource[T],
+        *,
+        ray_remote_args: Dict[str, Any] = None,
+        **write_args,
+    ) -> None:
+        """Call ``Dataset.write_datasource`` on each output dataset of this
+        pipeline."""
+        self._write_each_dataset(
+            lambda ds: ds.write_datasource(
+                datasource,
+                ray_remote_args=ray_remote_args,
+                **write_args,
+            )
+        )
+
+    def take(self, limit: int = 20) -> List[T]:
+        """Call ``Dataset.take`` over the stream of output batches from the
+        pipeline"""
+        return Dataset.take(self, limit)
```

Comment:
In a future PR, we should really factor out the shared `Dataset` and `DatasetPipeline` code into utilities, instead of binding a `DatasetPipeline` to a `Dataset` instance method and hoping that it works.

Reply:
+1 on this. I was surprised that Python allows to do it when I read it.